### PR TITLE
change(web): better fat-finger key weighting ✏️

### PR DESCRIPTION
**Pull request by jahorton**
Mitigates #7240 by more strongly weighting keys close to the touch point.

**Also** fixes interference that was being caused from 'hidden'-property keys, which resemble padding areas on the OSK.  If one of these was 'most likely', default suggestions would be displayed instead of suggestions for the most likely actual key.

**Additionally**, this fixes a bug with determining how far a touchpoint is from the center of the right-most key of each row of the on-screen-keyboard - this bug was causing overcorrections in favor of the key immediately to the left in most cases.

Previously, we were using an inverse-linear relationship for a key's probability based on its distance from the touchpoint; I now believe inverse-quadratic to be more appropriate.  In relation to #7201 (and [this comment](https://github.com/keymanapp/keyman/pull/7241/files#r967960102)), this helps make boundaries between keys "sharper" without interfering much with our ability to correct further down the line once more characters are added.

## User Testing

Run all user tests specified below on an Android device (or simulator).  I'm not picky about which.

**TEST_S_Q**:  Using the default (`sil_euro_latin`) keyboard, press the bottom edge of the 'q' (or 'Q') key.
- All suggestions should begin with the corresponding letter and case.  (So, 'question' for 'q' or 'Question' for 'Q'.)

**TEST_E_Q**:  Using the default (`sil_euro_latin`) keyboard, press the right edge of the 'q' (or 'Q') key.
- All suggestions should begin with the corresponding letter and case.  (So, 'question' for 'q' or 'Question' for 'Q'.)
- If you manage to land perfectly in the middle, you may see 'w' suggestions, but not 'a' suggestions.

**TEST_W_P**:  Using the default (`sil_euro_latin`) keyboard, press the far left edge of the 'p' (or 'P') key.
- All suggestions should begin with either an 'o' or a 'p'.
    - If you manage to land right near the edge, you may get a mix!
    - If you land a bit further right than that, you should get all 'p' suggestions.

(Note:  the precision required to land these near-key corrections may be better served with the "predictive text: robust testing" Web test page.  It's a good thing if they don't happen for you when testing via touch!)

**TEST_SW_Z**:  Using the default (`sil_euro_latin`) keyboard, press the far bottom-left edge of the 'z' (or 'Z') key.
- All suggestions should begin with with the corresponding letter and case.

**TEST_E_L_BLANK**:  Using the default (`sil_euro_latin`) keyboard, press past the right edge of the 'l' (or 'L') key.
- All suggestions should begin with with 'l' / 'L' if it is the closest key to your touchpoint.

**TEST_S_P_BLANK**:  Using the default (`sil_euro_latin`) keyboard, press past the bottom edge of the 'p' (or 'P') key.
- All suggestions should begin with with 'p' / 'P' if it is the closest key to your touchpoint, regardless of the text actually produced.

For the next test, you may wish to make note of the following resource:
- https://keyman.com/keyboards/sil_jarai?bcp47=jra-khmr
- In order to properly set things up for the test, please use the _**in-app keyboard search**_ to find and install this keyboard.
    - If you're missing predictive text after installing it, it's likely because you ignored the bolded part above.

**TEST_JARAI_SE_SHIFT_Q**:  Using the `sil_jarai` keyboard's shift layer, touch the far, far bottom-right corner of the first key in the second row.  (ឈ)
- All predictions should start with the corresponding letter.

### Other notes

Note that the fat-finger distribution is currently _not_ used by the web engine to determine the actually selected key.  As a few of the tests above highlight... it may be wise to actually 'fix' that, as there are some cases where the distance-measuring functionality definitely performs better.  That "S_P_BLANK" one is a particular standout in this regard.

### CODE REVIEW

**Comment by mcdurdin** [comment]
> In order to properly set things up for the test, please use the _**in-app keyboard search**_ to find and install this keyboard.
> * If you're missing predictive text after installing it, it's likely because you ignored the bolded part above.

Why is this necessary? Is this a bug in the web-based keyboard search route?

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts [hunk 1]
```diff
@@ -550,6 +555,8 @@ namespace com.keyman.keyboards {
             // Results in a more optimized distribution.
             if(text.Codes.isKnownOSKModifierKey(key.baseKeyID)) {
               return;
+            } else if(key.isPadding) { // to the user, blank / padding keys do not exist.
```
Comment (mcdurdin) [inline common/web/keyboard-processor/src/keyboards/activeLayout.ts:558]:
```suggestion
            } else if(key.isPadding) { // to the user, padding keys do not exist.
```

Per comment in `isPadding()` you are not testing for blank keys

File: common/web/input-processor/src/text/inputProcessor.ts [hunk 2]
```diff
@@ -273,6 +273,7 @@ namespace com.keyman.text {
           let totalMass = 0; // Tracks sum of non-error probabilities.
           for(let pair of keyDistribution) {
             if(pair.p < KEYSTROKE_EPSILON) {
+              totalMass += pair.p;
```
Comment (mcdurdin) [inline common/web/input-processor/src/text/inputProcessor.ts:276]: I'm not entirely sure which issue this is correcting?

**Review by mcdurdin — APPROVED** [review]

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts (hunk 1 above)
Comment (jahorton) [reply common/web/keyboard-processor/src/keyboards/activeLayout.ts:558]: Ah, right, blank vs hidden.

File: common/web/input-processor/src/text/inputProcessor.ts (hunk 2 above)
Comment (jahorton) [reply common/web/input-processor/src/text/inputProcessor.ts:276]: This helped with investigating the issue and debugging.  It allows me to infer the original key probability from the final prediction.

Without that, the probability was being changed and didn't match what I expected.

**Comment by jahorton** [comment]
> > In order to properly set things up for the test, please use the _**in-app keyboard search**_ to find and install this keyboard.
> > 
> > * If you're missing predictive text after installing it, it's likely because you ignored the bolded part above.
> 
> Why is this necessary? Is this a bug in the web-based keyboard search route?

I don't believe so.  It's just that if I were going to run the test myself, I could see myself downloading the package separately and then manually installing from the 'offline' file.  I'm not confident in _that_ path linking in predictive text.

**Comment by bharanidharanj** [comment]
* **TEST_S_Q (PASSED):** Tested this with the attached PR build (Keyman 16.0.61-alpha-test-7242) in API 26 / Android 8.0 emulator and pressing the bottom edge of the 'q' or 'Q' key produces the suggestions that appear with the corresponding letter and case. 

   

* **TEST_E_Q (PASSED):** Pressing the right edge of the 'q' or 'Q' key produces corresponding letter and case, also pressing the in the middle of the key produces 'w' suggestion on the banner.

   


* **TEST_W_P (PASSED):** Mixed suggestions appears when I press the right near edge of the key. If I click further right on the P key would show all 'P' suggestions.

   


* **TEST_SW_Z (FAILED):** Pressing the far bottom-left edge of the 'z' key produces suggestions that also begin with capital letter 'Z'

   


* **TEST_E_L_BLANK (PASSED):** Pressing 'l' or 'L' Key would shows all the suggestions that begin with the corresponding letter and case.

   


* **TEST_S_P_BLANK (PASSED):** Pressing the bottom edge of the 'p' or 'P' would show all the suggestion that begins with 'p' or 'P'. 

   


* **TEST_JARAI_SE_SHIFT_Q (FAILED):** Pressing ឈ key, does not begin with the corresponding letter.

**Comment by jahorton** [comment]
> * **TEST_SW_Z (FAILED):** Pressing the far bottom-left edge of the 'z' key produces suggestions that also begin with capital letter 'Z'
>    

**TEST_SW_Z: PASSED**  Zelda is an english name; names should always be capitalized.  (Notably, it's part of the name of a popular video game franchise.)  It still begins with the same letter.

> * **TEST_JARAI_SE_SHIFT_Q (FAILED):** Pressing ឈ key, does not begin with the corresponding letter.
>    

**TEST_JARAI_SE_SHIFT_Q:  PASSED**  The writing system got in the way a bit here; all suggestions do start with the specified letter; while the third one just has a vowel attached to its left-hand side, it's the base character that is "first" and that matters most here.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.67-alpha](https://keyman.com/downloads/releases/alpha/16.0.67)

**Pull request merged**